File: backend/app/routers/bulletin.py
"""
Company Bulletin API - a community feed (photo / link / text posts, likes,
comments). Deliberately lighter than the core app: no Sheets export, best-effort
posting, soft moderation. Everything is JWT-protected; removals are admin-only.
"""
from collections import defaultdict
from datetime import datetime
from typing import Optional

from fastapi import APIRouter, Depends, File, Form, HTTPException, Query, Response, UploadFile
from pydantic import BaseModel
from sqlalchemy import func
from sqlalchemy.exc import IntegrityError
from sqlalchemy.orm import Session

from app.core.deps import get_current_user, get_db, require_admin
from app.core.link_preview import fetch_link_preview
from app.db.models.bulletin import BulletinComment, BulletinLike, BulletinPost
from app.db.models.user import User

router = APIRouter(prefix="/api/bulletin", tags=["bulletin"])

FEED_PAGE = 20
MAX_TEXT = 5000


def _comment_out(c: BulletinComment) -> dict:
    return {
        "comment_uuid": c.comment_uuid,
        "author_id": c.author_id,
        "author_name": c.author_name,
        "text": c.text,
        "created_at": c.created_at.isoformat(),
    }


def _image_url(p: BulletinPost) -> Optional[str]:
    # Server-stored bytes are served from a capability URL keyed by post_uuid
    # (relative - the frontend prepends its API base). Legacy Drive posts fall
    # back to the Drive thumbnail (an absolute https URL).
    if p.image_mime:
        return f"/api/bulletin/image/{p.post_uuid}"
    return p.image_thumb_url


def _post_out(p: BulletinPost, like_count: int, liked: bool, comments: list) -> dict:
    return {
        "post_uuid": p.post_uuid,
        "author_id": p.author_id,
        "author_name": p.author_name,
        "kind": p.kind,
        "text": p.text,
        "image_url": _image_url(p),
        "image_thumb_url": None,
        "link_url": p.link_url,
        "link_title": p.link_title,
        "link_description": p.link_description,
        "link_image_url": p.link_image_url,
        "created_at": p.created_at.isoformat(),
        "like_count": like_count,
        "liked_by_me": liked,
        "comments": comments,
    }


@router.get("/latest")
def latest(
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    """The newest non-removed post id, for the nav "new activity" dot. Cheap
    enough to poll; the client compares it to the last id it has seen."""
    row = (
        db.query(BulletinPost.id)
        .filter(BulletinPost.removed_at.is_(None))
        .order_by(BulletinPost.id.desc())
        .first()
    )
    return {"latest_id": int(row[0]) if row else 0}


@router.get("/feed")
def feed(
    before_id: Optional[int] = Query(default=None),
    limit: int = Query(default=FEED_PAGE, ge=1, le=50),
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    """Newest-first page of non-removed posts, with like counts, whether the
    caller liked each, and the (non-removed) comments. `before_id` cursors older."""
    q = db.query(BulletinPost).filter(BulletinPost.removed_at.is_(None))
    if before_id is not None:
        q = q.filter(BulletinPost.id < before_id)
    posts = q.order_by(BulletinPost.id.desc()).limit(limit).all()
    if not posts:
        return {"posts": [], "next_before_id": None}

    ids = [p.id for p in posts]

    # Batch the likes + comments for the page (no per-post query).
    like_counts: dict[int, int] = dict(
        db.query(BulletinLike.post_id, func.count(BulletinLike.id))
        .filter(BulletinLike.post_id.in_(ids))
        .group_by(BulletinLike.post_id)
        .all()
    )
    my_likes = {
        pid for (pid,) in db.query(BulletinLike.post_id)
        .filter(BulletinLike.post_id.in_(ids), BulletinLike.user_id == current_user.id)
        .all()
    }
    comments_by_post: dict[int, list] = defaultdict(list)
    for c in (
        db.query(BulletinComment)
        .filter(BulletinComment.post_id.in_(ids), BulletinComment.removed_at.is_(None))
        .order_by(BulletinComment.id.asc())
        .all()
    ):
        comments_by_post[c.post_id].append(_comment_out(c))

    out = [
        _post_out(p, int(like_counts.get(p.id, 0)), p.id in my_likes, comments_by_post.get(p.id, []))
        for p in posts
    ]
    return {"posts": out, "next_before_id": ids[-1] if len(posts) == limit else None}


class PostIn(BaseModel):
    post_uuid: str
    kind: str  # "text" | "link"
    text: str = ""
    link_url: Optional[str] = None


@router.post("/posts")
def create_post(
    body: PostIn,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    """Create a text or link post. (Photo posts use /posts/photo.) Idempotent on
    post_uuid - a retry returns the existing post."""
    existing = db.query(BulletinPost).filter(BulletinPost.post_uuid == body.post_uuid).first()
    if existing:
        return _post_out(existing, 0, False, [])

    kind = body.kind if body.kind in ("text", "link") else "text"
    text = (body.text or "").strip()[:MAX_TEXT]

    link_url = link_title = link_desc = link_image = None
    if kind == "link":
        link_url = (body.link_url or "").strip()
        if not link_url:
            raise HTTPException(status_code=422, detail="A link post needs a URL.")
        preview = fetch_link_preview(link_url)
        link_title = preview.get("title")
        link_desc = preview.get("description")
        link_image = preview.get("image")
    elif not text:
        raise HTTPException(status_code=422, detail="A text post needs some text.")

    post = BulletinPost(
        post_uuid=body.post_uuid,
        author_id=current_user.id,
        author_name=(current_user.name or current_user.email or "").strip(),
        kind=kind,
        text=text,
        link_url=link_url,
        link_title=link_title,
        link_description=link_desc,
        link_image_url=link_image,
        created_at=datetime.utcnow(),
    )
    db.add(post)
    try:
        db.commit()
    except IntegrityError:
        db.rollback()  # raced the same post_uuid
        post = db.query(BulletinPost).filter(BulletinPost.post_uuid == body.post_uuid).first()
    return _post_out(post, 0, False, [])


MAX_IMAGE_BYTES = 8 * 1024 * 1024  # 8 MB. JPEGs are resized client-side first; GIFs upload raw to keep animation.


@router.post("/posts/photo")
def create_photo_post(
    file: UploadFile = File(...),
    post_uuid: str = Form(...),
    text: str = Form(default=""),
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    """Create a photo post. The image is stored server-side (bytes on the row) and
    served from a capability URL - no Drive, server-only and transient.
    Idempotent on post_uuid."""
    existing = db.query(BulletinPost).filter(BulletinPost.post_uuid == post_uuid).first()
    if existing:
        return _post_out(existing, 0, False, [])

    data = file.file.read(MAX_IMAGE_BYTES + 1)
    if not data:
        raise HTTPException(status_code=422, detail="Empty image.")
    if len(data) > MAX_IMAGE_BYTES:
        raise HTTPException(status_code=413, detail="Image too large - please try a smaller photo.")
    mime_type = file.content_type or "image/jpeg"
    if not mime_type.startswith("image/"):
        mime_type = "image/jpeg"

    post = BulletinPost(
        post_uuid=post_uuid,
        author_id=current_user.id,
        author_name=(current_user.name or current_user.email or "").strip(),
        kind="photo",
        text=(text or "").strip()[:MAX_TEXT],
        image_bytes=data,
        image_mime=mime_type,
        created_at=datetime.utcnow(),
    )
    db.add(post)
    try:
        db.commit()
    except IntegrityError:
        db.rollback()
        post = db.query(BulletinPost).filter(BulletinPost.post_uuid == post_uuid).first()
    return _post_out(post, 0, False, [])


@router.get("/image/{post_uuid}")
def get_image(post_uuid: str, db: Session = Depends(get_db)):
    """Serve a post's image bytes. Public (no auth header) because <img> tags
    can't send a bearer token; the random post_uuid is the capability, matching
    how job photos are public via unguessable Drive links."""
    p = db.query(BulletinPost).filter(BulletinPost.post_uuid == post_uuid).first()
    if not p or p.removed_at is not None or not p.image_bytes:
        raise HTTPException(status_code=404, detail="Image not found.")
    return Response(
        content=p.image_bytes,
        media_type=p.image_mime or "image/jpeg",
        headers={"Cache-Control": "public, max-age=86400"},
    )


def _require_post(db: Session, post_uuid: str) -> BulletinPost:
    p = db.query(BulletinPost).filter(BulletinPost.post_uuid == post_uuid).first()
    if not p or p.removed_at is not None:
        raise HTTPException(status_code=404, detail="Post not found.")
    return p


@router.post("/posts/{post_uuid}/like")
def toggle_like(
    post_uuid: str,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    """Toggle the caller's like on a post. Returns the new state + count."""
    post = _require_post(db, post_uuid)
    row = (
        db.query(BulletinLike)
        .filter(BulletinLike.post_id == post.id, BulletinLike.user_id == current_user.id)
        .first()
    )
    if row:
        db.delete(row)
        liked = False
    else:
        db.add(BulletinLike(post_id=post.id, user_id=current_user.id, created_at=datetime.utcnow()))
        liked = True
    try:
        db.commit()
    except IntegrityError:
        db.rollback()  # raced a concurrent like; treat as liked
        liked = True
    count = db.query(func.count(BulletinLike.id)).filter(BulletinLike.post_id == post.id).scalar() or 0
    return {"liked": liked, "like_count": int(count)}


class CommentIn(BaseModel):
    comment_uuid: str
    text: str = ""


@router.post("/posts/{post_uuid}/comments")
def add_comment(
    post_uuid: str,
    body: CommentIn,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    """Add a comment. Idempotent on comment_uuid."""
    post = _require_post(db, post_uuid)
    text = (body.text or "").strip()[:MAX_TEXT]
    if not text:
        raise HTTPException(status_code=422, detail="Empty comment.")
    existing = db.query(BulletinComment).filter(BulletinComment.comment_uuid == body.comment_uuid).first()
    if existing:
        return _comment_out(existing)
    c = BulletinComment(
        comment_uuid=body.comment_uuid,
        post_id=post.id,
        author_id=current_user.id,
        author_name=(current_user.name or current_user.email or "").strip(),
        text=text,
        created_at=datetime.utcnow(),
    )
    db.add(c)
    try:
        db.commit()
    except IntegrityError:
        db.rollback()
        c = db.query(BulletinComment).filter(BulletinComment.comment_uuid == body.comment_uuid).first()
    return _comment_out(c)


@router.delete("/posts/{post_uuid}")
def remove_post(
    post_uuid: str,
    db: Session = Depends(get_db),
    admin: User = Depends(require_admin),
):
    """Admin soft-removes a post (kept in the DB, hidden from the feed)."""
    p = db.query(BulletinPost).filter(BulletinPost.post_uuid == post_uuid).first()
    if not p:
        raise HTTPException(status_code=404, detail="Post not found.")
    if p.removed_at is None:
        p.removed_at = datetime.utcnow()
        p.removed_by = (admin.name or admin.email or "admin")
        db.commit()
    return {"ok": True}


@router.delete("/comments/{comment_uuid}")
def remove_comment(
    comment_uuid: str,
    db: Session = Depends(get_db),
    admin: User = Depends(require_admin),
):
    """Admin soft-removes a comment."""
    c = db.query(BulletinComment).filter(BulletinComment.comment_uuid == comment_uuid).first()
    if not c:
        raise HTTPException(status_code=404, detail="Comment not found.")
    if c.removed_at is None:
        c.removed_at = datetime.utcnow()
        c.removed_by = (admin.name or admin.email or "admin")
        db.commit()
    return {"ok": True}

File: frontend/src/lib/bulletin.ts
/**
 * Company Bulletin API client. A community feed, so it is intentionally lighter
 * than the offline-first core tools: posts go straight to the server with clear
 * errors (no localStorage queue), and the feed is fetch-on-mount + refresh. Each
 * post/comment carries a client-minted UUID so a double-tap can't duplicate.
 */
import { apiFetch } from "../api/client";
import { getToken } from "../auth/token";
import { coalesce } from "./sharedFetch";

const API = import.meta.env.VITE_API_URL || "http://127.0.0.1:8000";

export type BulletinComment = {
  comment_uuid: string;
  author_id: number | null;
  author_name: string;
  text: string;
  created_at: string;
};

export type BulletinPost = {
  post_uuid: string;
  author_id: number | null;
  author_name: string;
  kind: "photo" | "link" | "text";
  text: string;
  image_url: string | null;
  image_thumb_url: string | null;
  link_url: string | null;
  link_title: string | null;
  link_description: string | null;
  link_image_url: string | null;
  created_at: string;
  like_count: number;
  liked_by_me: boolean;
  comments: BulletinComment[];
  /** Which reaction this post accepts. The server decides; the client renders.
   *  Optional so a client running against an older backend still works - it
   *  falls back to "like", which is what every post was before this existed. */
  reaction_mode?: ReactionMode;
  /** Whether THIS viewer may switch the mode. Computed server-side; the client
   *  never holds the rule about who is allowed. Hiding the control is a
   *  courtesy - the endpoint refuses anyone else regardless. */
  can_set_reaction_mode?: boolean;
};

export type ReactionMode = "like" | "dislike";

export type Feed = { posts: BulletinPost[]; next_before_id: number | null };

function uuid(): string {
  return crypto.randomUUID ? crypto.randomUUID() : `${Date.now()}-${Math.random().toString(36).slice(2)}`;
}

/**
 * Coerce one post into a shape the UI can render.
 *
 * `apiFetch<BulletinPost>` is a TYPE assertion, not a runtime check - the
 * generic is erased, so a degraded server response flows straight into JSX. A
 * post arriving without `comments` blows up on `post.comments.length` during
 * render, which unmounts the tree behind the app-wide ErrorBoundary. Guaranteeing
 * the array here is cheaper than guarding every read of it.
 */
function normalizePost(raw: any): BulletinPost {
  return {
    ...(raw as BulletinPost),
    comments: Array.isArray(raw?.comments) ? raw.comments : [],
  };
}

export async function fetchFeed(beforeId?: number | null): Promise<Feed> {
  const q = beforeId ? `?before_id=${beforeId}` : "";
  const raw = await apiFetch<any>(`/api/bulletin/feed${q}`);
  // A malformed body must be an ERROR, not an empty feed. Without this, a
  // degraded response (a 200 carrying an error object, a proxy page, a partial
  // payload) renders as "Nothing posted yet" - which tells a crew member the
  // bulletin is empty when it is actually broken, and gives them no Retry.
  if (!raw || !Array.isArray(raw.posts)) {
    throw new Error("The bulletin sent back something this app could not read.");
  }
  return {
    posts: raw.posts
      .filter((p: any) => p && typeof p === "object")
      .map(normalizePost),
    next_before_id:
      typeof raw.next_before_id === "number" ? raw.next_before_id : null,
  };
}

export function createTextPost(text: string): Promise<BulletinPost> {
  return apiFetch<any>("/api/bulletin/posts", {
    method: "POST",
    body: JSON.stringify({ post_uuid: uuid(), kind: "text", text }),
  }).then(normalizePost);
}

export function createLinkPost(linkUrl: string, text: string): Promise<BulletinPost> {
  return apiFetch<any>("/api/bulletin/posts", {
    method: "POST",
    body: JSON.stringify({ post_uuid: uuid(), kind: "link", link_url: linkUrl, text }),
  }).then(normalizePost);
}

// Resize a photo to a sane max dimension + JPEG before upload, so the bytes we
// store server-side (and load into the worker to serve) stay small. Falls back
// to the original file if the canvas path fails.
async function resizeImage(file: File, maxPx = 1600, quality = 0.85): Promise<Blob> {
  try {
    const bmpUrl = URL.createObjectURL(file);
    const img = await new Promise<HTMLImageElement>((resolve, reject) => {
      const el = new Image();
      el.onload = () => resolve(el);
      el.onerror = reject;
      el.src = bmpUrl;
    });
    const scale = Math.min(1, maxPx / Math.max(img.width, img.height));
    const w = Math.round(img.width * scale);
    const h = Math.round(img.height * scale);
    const canvas = document.createElement("canvas");
    canvas.width = w; canvas.height = h;
    const ctx = canvas.getContext("2d");
    if (!ctx) throw new Error("no canvas");
    ctx.drawImage(img, 0, 0, w, h);
    URL.revokeObjectURL(bmpUrl);
    const blob = await new Promise<Blob | null>((res) => canvas.toBlob(res, "image/jpeg", quality));
    return blob || file;
  } catch {
    return file;
  }
}

export async function createPhotoPost(file: File, text: string): Promise<BulletinPost> {
  // GIFs (usually animated) must NOT go through the canvas resize - that flattens
  // them to a single JPEG frame and kills the animation. Send the original bytes;
  // everything else is resized/compressed to keep the stored image small.
  const upload =
    file.type === "image/gif"
      ? file
      : new File([await resizeImage(file)], "photo.jpg", { type: "image/jpeg" });
  const form = new FormData();
  form.append("file", upload);
  form.append("post_uuid", uuid());
  form.append("text", text);
  const res = await fetch(`${API}/api/bulletin/posts/photo`, {
    method: "POST",
    headers: { Authorization: `Bearer ${getToken() || ""}` },
    body: form,
  });
  const json = await res.json().catch(() => ({}));
  if (!res.ok) throw new Error(json?.detail || `Upload failed (HTTP ${res.status})`);
  // Same guarantee as the feed: a post reaching component state without a
  // `comments` array crashes the render that reads `post.comments.length`.
  return normalizePost(json);
}

// image_url is relative for server-stored images ("/api/bulletin/image/...") and
// absolute for legacy Drive posts; resolve both to a usable <img src>.
export function postImageSrc(post: BulletinPost): string | null {
  const u = post.image_url;
  if (!u) return null;
  return u.startsWith("http") ? u : `${API}${u}`;
}

// ── "New activity" dot for the nav ──
const SEEN_KEY = "crew_bulletin_seen_id_v1";
export function getSeenId(): number {
  try { return Number(localStorage.getItem(SEEN_KEY)) || 0; } catch { return 0; }
}
export function setSeenId(id: number): void {
  try { localStorage.setItem(SEEN_KEY, String(id)); } catch { /* noop */ }
}
export function fetchLatestId(): Promise<{ latest_id: number }> {
  // Called by the bottom nav on EVERY navigation to decide whether to show a
  // dot, which made this 11% of all backend traffic in a production sample. A
  // dot saying "somebody posted" does not need per-tap freshness, and every
  // request spent here brings the 1000-request worker recycle closer.
  return coalesce(
    "bulletin:latest",
    () => apiFetch<{ latest_id: number }>("/api/bulletin/latest"),
    { ttlMs: 60_000 },
  );
}

export function toggleLike(postUuid: string): Promise<{ liked: boolean; like_count: number }> {
  return apiFetch(`/api/bulletin/posts/${encodeURIComponent(postUuid)}/like`, { method: "POST" });
}

/**
 * Switch one post between accepting likes and accepting dislikes.
 *
 * Returns 404 (not 403) for anyone who is not permitted, so a caller that has no
 * business knowing this endpoint exists learns nothing from it.
 */
export function setReactionMode(
  postUuid: string,
  mode: ReactionMode,
): Promise<{ post_uuid: string; reaction_mode: ReactionMode; like_count: number }> {
  return apiFetch(`/api/bulletin/posts/${encodeURIComponent(postUuid)}/reaction-mode`, {
    method: "POST",
    body: JSON.stringify({ mode }),
  });
}

export function addComment(postUuid: string, text: string): Promise<BulletinComment> {
  return apiFetch<BulletinComment>(`/api/bulletin/posts/${encodeURIComponent(postUuid)}/comments`, {
    method: "POST",
    body: JSON.stringify({ comment_uuid: uuid(), text }),
  });
}

export function removePost(postUuid: string): Promise<{ ok: boolean }> {
  return apiFetch(`/api/bulletin/posts/${encodeURIComponent(postUuid)}`, { method: "DELETE" });
}

export function removeComment(commentUuid: string): Promise<{ ok: boolean }> {
  return apiFetch(`/api/bulletin/comments/${encodeURIComponent(commentUuid)}`, { method: "DELETE" });
}

// "3m", "2h", "4d", or a date for older.
export function timeAgo(iso: string): string {
  const then = new Date(iso).getTime();
  const secs = Math.max(0, Math.floor((Date.now() - then) / 1000));
  if (secs < 60) return "just now";
  const mins = Math.floor(secs / 60);
  if (mins < 60) return `${mins}m`;
  const hrs = Math.floor(mins / 60);
  if (hrs < 24) return `${hrs}h`;
  const days = Math.floor(hrs / 24);
  if (days < 7) return `${days}d`;
  return new Date(iso).toLocaleDateString([], { month: "short", day: "numeric" });
}
